stages: add new unit test for kickstart stage

mvo5 opened the pull request:
This commit adds a simple and lightweight unit test for the new kickstart options. It's pretty trivial but also cheap and runs fast. The idea behind it is that the code of the kickstart stage is tested without the need for a full osbuild tree, just passing the test input options and observing the expected file content. The same pattern could potentially be used for other stages (like the user stage) to check that the content of the output has the expected format (or that options that are mutually exclusive are handled correctly). 

While this is trivial now and looks a bit like a waste of time I think this will become useful as more kickstart stage options get implemented. The goal is to eventually support the options in https://github.com/RedHatInsights/edge-api/blob/main/templates/templateKickstart.ks in a declarative manner and part of this will involve supporting more things that kickstart is currently doing, e.g https://github.com/RedHatInsights/edge-api/blob/main/templates/templateKickstart.ks#L77 and here we will need to be careful about things like input validation/escaping so having this unit tested seems quite useful. I personally also like it because it tells me at a glance what the kickstart file should look like (test as docuementation).

Note that this is not urgent at all and given how trivial it is I'm fine with being told that this is over-testing. We could keep it in draft stage (or close it even) until there is a more complex use case (as stated above I anticipate there will be one for kickstart).

I'm not happy with chage chage in line 22 (tox.ini), it's needed because with the "*" glob pylint will consider the test dir a python module but because it has no __init__.py that fails. Adding a __init__.py will ake pytest fail. The alternative would be to move this under the top level `test/` dir (e.g. under `test/unit/stages` or something.

Having said all this - I would love your intput and feedback :)

supakeen commented on stages/test/org_osbuild_kickstart.py:
```diff
@@ -0,0 +1 @@
+../org.osbuild.kickstart
```
I'd prefer not to have a symlink under a different name, see later comment about `importlib`.

supakeen commented on stages/test/test_kickstart.py:
```diff
@@ -0,0 +1,53 @@
+import unittest
+from tempfile import mkdtemp
+
+from org_osbuild_kickstart import main as kickstart_main
```
We could do this by using `importlib` instead; note that it's a bit more code that we might want to share over multiple stage test cases and we might run into some issues compatibility wise (some things have changed over time in `importlib`) so I'm not going to press hard on this. It'd just be nicer :)

mvo5 replied:
Thanks! Indeed, my first impulse was importlib too, then I started reading the docs and got a bit discouraged and went with the symlink to illustrate my idea. But for the real thing(tm) using something nicer makes a ton of sense :)

supakeen commented on stages/test/test_kickstart.py:
```diff
@@ -0,0 +1,53 @@
+class TestOrgOsbuilderKickstart(unittest.TestCase):
+
+    def setUp(self):
+        self.tmpdir = mkdtemp()
```
This can be done with `with tempfile.TemporaryDirectory` which I kinda like.

mvo5 replied:
I went with the pytest tmp_path fixture here now

supakeen commented on stages/test/test_kickstart.py:
```diff
@@ -0,0 +1,53 @@
+
+    def test_kickstart(self):
+        ks_path = "kickstart/kfs.cfg"
+        for tc in test_cases:
```
Let's use pytest fixtures for this instead we're already using pytest as the test runner :)

mvo5 replied:
Sure thing, I "rewrote" it and went pytest all the way now (hope that was okay)

supakeen commented:
This is neither a waste of time nor trivial. I'd *love* to have a quicker way to test stages doing the right things :)

supakeen commented:
> I'm not happy with chage chage in line 22 (tox.ini), it's needed because with the "*" glob pylint will consider the test dir a python module but because it has no **init**.py that fails. Adding a **init**.py will ake pytest fail. The alternative would be to move this under the top level `test/` dir (e.g. under `test/unit/stages` or something.

I like the stage tests living close to the stages. People do write their own stages as well and it fits better in future (tm) plans to support different types of stages which each have their own test.

mvo5 commented on stages/test/test_kickstart.py:
```diff
@@ -0,0 +1,56 @@
+
+
+@dataclass
+class KsTestCase:
```
Not sure if this is overdoing it TBH but the alternative would be just using a dict

supakeen replied:
Due to our `>= 3.6` constraint for Python we can't use dataclasses :(

mvo5 commented:
> > I'm not happy with chage chage in line 22 (tox.ini), it's needed because with the "*" glob pylint will consider the test dir a python module but because it has no **init**.py that fails. Adding a **init**.py will ake pytest fail. The alternative would be to move this under the top level `test/` dir (e.g. under `test/unit/stages` or something.
> 
> I like the stage tests living close to the stages. People do write their own stages as well and it fits better in future (tm) plans to support different types of stages which each have their own test.

Thanks, for your excellent feedback! I addressed the points and moved the PR from draft to real and tried to polish it. I'm sure there are a bunch of fixes needed around exact naming and location of the files etc but I hope the shape is now looking good. Fwiw, I'm happy to split the import helper out but given that it's small and the test is the use-case I felt like it might be a good idea to just leave it in :)

supakeen commented on stages/test/test_kickstart.py:
```diff
@@ -0,0 +1,55 @@
+    expected: str
+
+
+@pytest.fixture(name="ks_test_cases")
```
Let's do this as a `@pytest.mark.parametrize`: https://docs.pytest.org/en/7.3.x/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions instead. We can name the list if we plan to use it for multiple tests :)

mvo5 replied:
Thank you, this is excellent! I updated the test and it's indeed much nicer now

supakeen reviewed (dismissed):
Love it. Let's keep it out of CI for now so we can work with it locally for a bit then enable it later or do we want it in now? (cc: @achilleas-k).

mvo5 commented:
> Love it. Let's keep it out of CI for now so we can work with it locally for a bit then enable it later or do we want it in now? (cc: @achilleas-k).

Thank you for the review and approval! 

I'm not sure I understand what `keep it out of CI` means? Maybe I'm misunderstanding but I would love to see it merged so that I can add test cases to PR#1426 :)

mvo5 commented on .github/workflows/test.yml:
```diff
@@ -19,6 +19,7 @@ jobs:
         - "test.run.test_noop"
         - "test.run.test_sources"
         - "test.run.test_stages"
+        - "-k stages/test"
```
Note that because `stages/__init__.py` and `stages/test/__init__.py` are missing it is not possible to use the existing style of just doing `stages.test`. Adding `stages/__init__.py` feels wrong and the desire is to have the stages tests close to the stages so this seems the least invasive way.

But happy to reconsider/adjust if you prefer something else.

supakeen replied:
This is ok :)

achilleas-k commented:
Does putting a `test/` directory under `stages/` affect packaging?  I think our spec file packages everything under that directory.  Do we want these new stage tests distributed in the rpms??

mvo5 commented:
> Does putting a `test/` directory under `stages/` affect packaging? I think our spec file packages everything under that directory. Do we want these new stage tests distributed in the rpms??

Thanks, excellent catch (that would have been slightly embarrassing to release with that)! I updated the spec file now to exclude anything `test_*.py`. Or if you feel the amount of workarounds get unwieldy we could also move into a separate `test/stages/unit` dir (but of course the downside of that is that tests/code are no longer as close as before).

supakeen reviewed (approved):
Very nice, can't wait to expand on this :)

achilleas-k reviewed (approved):
LGTM
